src/chromatix/functional/propagation.py

```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
     delay = jnp.sqrt(jnp.abs(kernel))
     delay = jnp.where(kernel >= 0, delay, 1j * delay)  # keep evanescent modes
     phase = 2 * jnp.pi * (z * n / field.spectrum) * delay
-    return jnp.fft.ifftshift(jnp.exp(1j * phase), axes=field.spatial_dims)
+    kernel_field = jnp.exp(1j * phase)
+
+    if bandlimit:
+        Sy, Sx = (1 / field.dk).squeeze()   # spatial dimension in microns
+        y0, x0 = (kykx / field.dk).squeeze()  # spatial shift in microns, TODO check
+        z0 = z.squeeze()  # propagation distance in microns
+        wv = field.spectrum.squeeze()  # wavelength in microns
+
+        dfX = 1.0 / Sx
+        dfY = 1.0 / Sy
+        N_y, N_x = field.spatial_shape
+        fX = np.linspace(-N_x // 2, N_x // 2 - 1, num=N_x)[np.newaxis, :] * dfX
+        fY = np.linspace(-N_y // 2, N_y // 2 - 1, num=N_y)[:, np.newaxis] * dfY
+
+        # Table 1 of "Shifted angular spectrum method for off-axis numerical 
+        # propagation" (2010) by Matsushima
+        du = 1 / (2 * Sx)
+        u_limit_p = ((x0 + 1 / (2 * du)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        u_limit_n = ((x0 - 1 / (2 * du)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        if Sx < x0:
+            u0 = (u_limit_p + u_limit_n) / 2
+            u_width = u_limit_p - u_limit_n
+        elif x0 <= -Sx:
+            u0 = -(u_limit_p + u_limit_n) / 2
+            u_width = u_limit_n - u_limit_p
+        else:
+            u0 = (u_limit_p - u_limit_n) / 2
+            u_width = u_limit_p + u_limit_n
```

Comment:
We have to be careful here - jax _really_ doesn't like conditionals based on values, so we shouldn't do this. Is there some way to write this without if/else?

Reply:
Maybe using `jnp.sign(x0 - Sx)` or something like that?

Reply:
yep I think applying `jnp.sign` correctly does the trick